[NativeAOT-LLVM] WASI fixes

SingleAccretion opened the pull request:
Following up on the thread in #2235.

Fixes #2385.
Fixes #2384.

Also upgrades the "missing SDKs" handling to errors and adds a test for a WASI reactor using NodeJS.

SingleAccretion committed:
Fix a number of issues with WASI libraries

1) Write and consume the exports list
2) Do not call "_initialize" twice.
3) Add "-mexec-model=reactor" to the linker command line

SingleAccretion committed:
Add a test

SingleAccretion committed:
Error out on missing SDKs

RReverser commented on src/coreclr/nativeaot/Bootstrap/main.cpp:
```diff
@@ -181,13 +182,20 @@ __attribute__((weak)) void _initialize()
 {
 #if defined(NATIVEAOT_DLL) && defined(TARGET_WASI)
-    _initialize();
+    if (!g_CalledInitialize)
```
Perhaps it would be simpler to have `static bool alreadyInitialized = false;` inside this function instead of separate struct that relies on static constructor? (assuming this function is called from inside `_initialize`)

SingleAccretion replied:
> assuming this function is called from inside _initialize

It isn't - it's called by the runtime on thread attach. So we need something from inside `_initialize` itself to notify us it has already run. Explicit static constructor is then that something.

RReverser replied:
Ah ok, sorry.

SingleAccretion replied:
No problem, thank you for the review!

RReverser commented on src/tests/nativeaot/SmokeTests/SharedLibrary/SharedLibraryDriver.mjs:
```diff
@@ -0,0 +1,39 @@
+// if (instance.exports.CheckSimpleGCCollect() != 100)
+//     process.exit(4);
+
+// #if !CODEGEN_WASI - enable when we support exception handling
```
I actually tried some simple exception handling examples and somehow they worked, but in some other scenarios it didn't. I'm not sure what it depends on - LLVM optimisations?

RReverser replied:
For example, I had something like this in my sample and it compiled and ran fine:

```cs
	[UnmanagedCallersOnly(EntryPoint = "hello")]
	public static void Hello() {
		try {
			Console.WriteLine("Hello from exception");
		} catch (Exception e) {
			Console.WriteLine(e);
		}
		Console.WriteLine("Hello from C#");
		MyMethod("lores ipsum");
	}
```

SingleAccretion replied:
What did and what did not work?

There is a dependency on `emscripten_get_callstack` inside exception handling (it's called on each `throw`), so I would _expect_ things to fail relatively deterministically.

Edit: that example doesn't `throw`, so I suppose it is expected to work.

RReverser replied:
> Edit: that example doesn't `throw`, so I suppose it is expected to work.

Oh my god, this is ridiculous. I need sleep 😬

SingleAccretion committed:
Fully specify the target triplet

Avoids warnings such as the following:

EXEC : warning : overriding the module target triple with wasm32-unknown-wasi [-Woverride-module]

SingleAccretion committed:
Up node the NodeJS version for CI

v18.16 is missing the WASI package.

SingleAccretion commented:
@dotnet/nativeaot-llvm

SingleAccretion commented on src/tests/nativeaot/SmokeTests/SharedLibrary/SharedLibraryDriver.mjs:
```diff
@@ -0,0 +1,39 @@
+// As long as no unmanaged exception is thrown managed class loaders were initialized successfully.
+instance.exports.EnsureManagedClassLoaders();
+
+// #if !CODEGEN_WASI - for some reason tries to create a background thread?
+// if (instance.exports.CheckSimpleGCCollect() != 100)
+//     process.exit(4);
```
I have the identified the root cause for this and will PR the fix separately.

yowl commented on src/coreclr/nativeaot/Bootstrap/main.cpp:
```diff
@@ -168,11 +168,12 @@ extern "C" int __managed__Main(int argc, char* argv[]);
 #define NATIVEAOT_ENTRYPOINT __managed__Startup
 extern "C" void __managed__Startup();
+
+#ifdef TARGET_WASI
```
Why is `NATIVEAOT_DLL` dropped here but still included in `InitializeRuntime` ?   If this is intended then I suppose the comment on 175 should be changed?

SingleAccretion replied:
It's not really dropped, it just wasn't needed, since we're in the `#else` part of `#ifndef NATIVEAOT_DLL`:
```cpp
#ifndef NATIVEAOT_DLL
#define NATIVEAOT_ENTRYPOINT __managed__Main
#if defined(_WIN32)
extern "C" int __managed__Main(int argc, wchar_t* argv[]);
#else
extern "C" int __managed__Main(int argc, char* argv[]);
#endif
#else // NATIVEAOT_DLL guaranteed to be defined below
#define NATIVEAOT_ENTRYPOINT __managed__Startup
extern "C" void __managed__Startup();

#ifdef TARGET_WASI
```

yowl reviewed (commented):
Thanks, just a question on `NATIVEAOT_DLL`

jkotas commented on src/coreclr/nativeaot/Bootstrap/main.cpp:
```diff
@@ -181,13 +182,20 @@ __attribute__((weak)) void _initialize()
 }
 
+// Guard the "_initialize" call so that well-behaving hosts do not get affected by this workaround.
+bool g_CalledInitialize = false;
```
```suggestion
static bool g_CalledInitialize = false;
```
?

jkotas reviewed (approved):
Thank you!

SingleAccretion committed:
Make the flag TU-local

Co-authored-by: Jan Kotas <jkotas@microsoft.com>
